contracts/upgradeability/ClassicEternalStorageProxy.sol

```diff
@@ -0,0 +1,35 @@
+pragma solidity 0.4.19;
+
+import "./EternalStorage.sol";
+import "./OwnedUpgradeabilityProxy.sol";
+
+
+contract ClassicEternalStorageProxy is OwnedUpgradeabilityProxy, EternalStorage {
+
+    function () payable public {
+        address _impl = implementation();
+        require(_impl != address(0));
+        bytes4 sig;
+        assembly { sig := calldataload(0) }
+        uint256 len = getSize(sig);
+        assembly {
+            calldatacopy(0x0, 0x0, calldatasize)
+            let result := delegatecall(gas, _impl, 0x0, calldatasize, 0, len)
+
+            switch result
+            case 0 { revert(0, len) }
+            default { return(0, len) }
+        }
+    }
+
+    function getSize(bytes4 sig) public pure returns(uint256) {
```

Comment:
Consider to create a separate file `ReturnDataSizes.sol` with the following content:
```solidity
contract ReturnDataSizesRegistrar {
    mapping(bytes4 => uint256) internal returnDataSizes;
}
```

so, the `ClassicEternalStorageProxy` contract and its `getSize()` method could look like:
```solidity
import "./OwnedUpgradeabilityProxy.sol";
import "./ReturnDataSizes.sol";

contract ClassicEternalStorageProxy is ReturnDataSizesRegistrar, OwnedUpgradeabilityProxy, EternalStorage {
    . . .
    . . .
    function getSize(bytes4 sig) public pure returns(uint256) {
        uint256 ret = returnDataSizes[sig];
        if (ret == 0) {
            ret = 32;
        }
        return ret;
    }
```

In this case `BasicHomeBridge.sol` will be changed like this:
```solidity
import "../libraries/Message.sol
import "../upgradeability/ReturnDataSizesRegistrar.sol";

contract BasicHomeBridge is ReturnDataSizesRegistrar, EternalStorage, Validatable {
    using SafeMath for uint256;
    returnDataSizes[bytes4(sha3("signature(bytes32,uint256)")] = 132;
    returnDataSizes[bytes4(sha3("message(bytes32)")] = 210;
```

Reply:
This will provide more logical way to define returned data sizes in the modules where the corresponding method is defined. And will reduce a bit the size of other contracts using the proxy template (e.g. the Validator contract).

Reply:
And the question is if it is safe enough to use `returnDataSizes`? Does it make sense to use `uintStorage`?

Reply:
Perhaps using the existing EternalStorage, in this case `uintStorage` could be a safer idea

Reply:
I like the rest of the approach, the only thing is that I think we should initialize the variables on `initialize` method of `HomeBridge`
```
returnDataSizes[bytes4(sha3("signature(bytes32,uint256)"))] = 132;
returnDataSizes[bytes4(sha3("message(bytes32)"))] = 210;
```
Otherwise they won't be available as part of the state on the proxy contract

Reply:
Correct. Thanks for pointing this out!

Reply:
Implemented the changes here: https://github.com/poanetwork/poa-bridge-contracts/pull/161/commits/a072da6e89a00979c052c1d31aab40d9d1657e45